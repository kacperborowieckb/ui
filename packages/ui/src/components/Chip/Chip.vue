<template>
  <div
    :tabindex="isFocusable ? 0 : undefined"
    :class="chipVariants({ variant })"
    @keydown.backspace="handleChipRemoval"
  >
    <span class="flex gap-2 py-[3px] leading-none">
      <template v-if="content">
        {{ content }}
      </template>
      <slot />
    </span>
    <button
      v-if="withButton"
      tabindex="-1"
      class="cursor-pointer"
      @click="handleChipRemoval"
    >
      <Cross2Icon />
    </button>
  </div>
</template>

<script lang="ts">
import { Cross2Icon } from '@radix-icons/vue'
import { tv } from 'tailwind-variants'
import { computed } from 'vue'
import type { VariantProps } from 'tailwind-variants'

const chipVariants = tv({
  base: 'focus-visible:ring-offset-background focus-visible:ring-ring flex items-center gap-1 rounded-sm border px-2 transition-colors focus-visible:outline-none focus-visible:ring-2 ',
  variants: {
    variant: {
      default: 'bg-muted dark:bg-muted/80 dark:hover:bg-input text-secondary-foreground hover:bg-input/75 border-input',
      error: 'bg-destructive/10 border-destructive/20 text-destructive hover:bg-destructive/20 dark:brightness-150',
    },
  },
})

type ChipVariants = VariantProps<typeof chipVariants>

export interface ChipProps {
  content?: string
  variant?: ChipVariants['variant']
  focusable?: boolean
  disabled?: boolean
  withButton?: boolean
}

export interface ChipEmits {
  (e: 'chipRemoval'): void
}
</script>

<script setup lang="ts">
const props = withDefaults(defineProps<ChipProps>(), {
  variant: 'default',
})

const emit = defineEmits<ChipEmits>()

const isFocusable = computed(() => props.focusable && !props.disabled)

function handleChipRemoval() {
  emit('chipRemoval')
}
</script>
