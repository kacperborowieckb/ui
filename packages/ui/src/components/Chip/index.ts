export {
  default as Chip,
  type ChipEmits,
  type ChipProps,
} from './Chip.vue'
