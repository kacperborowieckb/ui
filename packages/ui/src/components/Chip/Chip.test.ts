import { shallowMount } from '@vue/test-utils'

import Chip from './Chip.vue'

describe('chip', () => {
  it('should render correctly without button', () => {
    const wrapper = shallowMount(Chip, {
      props: {
        content: 'mockContent',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with button', () => {
    const wrapper = shallowMount(Chip, {
      props: {
        withButton: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with error state', () => {
    const wrapper = shallowMount(Chip, {
      props: {
        withButton: true,
        content: 'mockContent',
        variant: 'error',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have tabindex set to 0 if focusable', () => {
    const wrapper = shallowMount(Chip, { props: { focusable: true } })

    expect(wrapper.attributes('tabindex')).toBe('0')
  })

  it('should not have tabindex if disabled', () => {
    const wrapper = shallowMount(Chip, { props: { focusable: true, disabled: true } })

    expect(wrapper.attributes('tabindex')).toBeUndefined()
  })

  it('should not have tabindex by default', () => {
    const wrapper = shallowMount(Chip)

    expect(wrapper.attributes('tabindex')).toBeUndefined()
  })

  it('should not emit chipRemoval on not backspace click', () => {
    const wrapper = shallowMount(Chip)

    wrapper.trigger('keydown.alt')

    expect(wrapper.emitted('chipRemoval')).toBeUndefined()
  })

  it('should emit chipRemoval event on backspace', () => {
    const wrapper = shallowMount(Chip)

    wrapper.trigger('keydown.backspace')

    expect(wrapper.emitted('chipRemoval')).toBeDefined()
  })

  it('should emit chipRemoval event on button click', () => {
    const wrapper = shallowMount(Chip, { props: { withButton: true } })

    const button = wrapper.find('button')

    button.trigger('click')

    expect(wrapper.emitted('chipRemoval')).toBeDefined()
  })
})
