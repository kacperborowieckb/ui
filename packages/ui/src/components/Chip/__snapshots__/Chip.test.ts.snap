// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chip > should render correctly with button 1`] = `
"<div class="focus-visible:ring-offset-background focus-visible:ring-ring flex items-center gap-1 rounded-sm border px-2 transition-colors focus-visible:outline-none focus-visible:ring-2 bg-muted dark:bg-muted/80 dark:hover:bg-input text-secondary-foreground hover:bg-input/75 border-input"><span class="flex gap-2 py-[3px] leading-none"><!--v-if--></span><button tabindex="-1" class="cursor-pointer">
    <render-stub></render-stub>
  </button></div>"
`;

exports[`chip > should render correctly with error state 1`] = `
"<div class="focus-visible:ring-offset-background focus-visible:ring-ring flex items-center gap-1 rounded-sm border px-2 transition-colors focus-visible:outline-none focus-visible:ring-2 bg-destructive/10 border-destructive/20 text-destructive hover:bg-destructive/20 dark:brightness-150"><span class="flex gap-2 py-[3px] leading-none">mockContent</span><button tabindex="-1" class="cursor-pointer">
    <render-stub></render-stub>
  </button></div>"
`;

exports[`chip > should render correctly without button 1`] = `
"<div class="focus-visible:ring-offset-background focus-visible:ring-ring flex items-center gap-1 rounded-sm border px-2 transition-colors focus-visible:outline-none focus-visible:ring-2 bg-muted dark:bg-muted/80 dark:hover:bg-input text-secondary-foreground hover:bg-input/75 border-input"><span class="flex gap-2 py-[3px] leading-none">mockContent</span>
  <!--v-if-->
</div>"
`;
